### PR TITLE
SONARJAVA-3759 Rule S6212: Local-Variable Type Inference (var) should…

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**
… be used

### CODE REVIEW

File: its/ruling/src/test/resources/eclipse-jetty/java-S6212.json [hunk 1]
```diff
@@ -0,0 +1,4066 @@
+{
```
Comment (margarita-nedzelska-sonarsource) [inline its/ruling/src/test/resources/eclipse-jetty/java-S6212.json:1]: I checked nearly 500 issues. The majority of them where TP, some of them where FP to me, so I updated the rule implementation with some constraints, where it's not possible/not recommended to use "var"

File: java-checks/src/test/java/org/sonar/java/checks/VarCanBeUsedCheckTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.sonar.java.checks.verifier.JavaCheckVerifier;
+
+import static org.sonar.java.checks.verifier.TestUtils.testSourcesPath;
+
+class VarCanBeUsedCheckTest {
+
+  @Test
+  void test() {
+    JavaCheckVerifier.newVerifier()
+      .onFile(testSourcesPath("checks/VarCanBeUsedCheck.java"))
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/VarCanBeUsedCheckTest.java:32]: You can probably extract `testSourcesPath("checks/VarCanBeUsedCheck.java")` into a field.

File: java-checks-test-sources/src/main/java/checks/VarCanBeUsedCheck.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+class MyTestClass {}
+
+public class VarCanBeUsedCheck {
+
+  public static final MyTestClass MyTest = new MyTestClass(); // Compliant
+  
+
```
Comment (quentin-jaquier-sonarsource) [inline java-checks-test-sources/src/main/java/checks/VarCanBeUsedCheck.java:12]: Extra empty lines.

File: java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      return false;
+    }
+    int line = firstToken.line();
+    if (typeAssignmentLine == line) {
+      return true;
+    }
+    typeAssignmentLine = line;
+    SyntaxToken token = variableTree.endToken();
+    return token != null && token.text().equals(",");
+  }
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java:94]: This code looks strange but I'm not sure there is another way to write it... So I guess we will keep it as it is.

**Review by quentin-jaquier-sonarsource — COMMENTED** [review]
The implementation LGTM. Let's wait the outcome of the discussion about the severity/profile before merging.

File: java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java (hunk 4 above)
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java:95]: I tried to find how it's handled in other rules and it was a similar approach. In situations like this `int a = 0, b = 1;` we can't put `var`, but there is no dedicated Kind. Maybe we should have it in the future.

**Comment by margarita-nedzelska-sonarsource** [comment]
I decided to filter out primitives and parameterized types at the moment. The rule still suggests a lot. Let's see how it goes and probably we can increase the scope in the future.

**Review by quentin-jaquier-sonarsource — COMMENTED** [review]
Almost there, let's finish this rule!

File: java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      !JUtils.isLocalVariable(variableTree.symbol()) ||
+      initializer.is(Tree.Kind.CONDITIONAL_EXPRESSION) ||
+      symbolType.isPrimitive() ||
+      symbolType.isParameterized()) {
+      return;
+    }
+
+    Type initializerType = initializer.symbolType();
+    if (symbolType.fullyQualifiedName().equals(initializerType.fullyQualifiedName())) {
+      reportIssue(variableTree.simpleName(), "");
```
[the remark this answers is not in the document]
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java:79]: Ooops

File: java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    if (symbolType.fullyQualifiedName().equals(initializerType.fullyQualifiedName()) &&
+      (isSelfExplained(initializer) ||
+        typeWasMentionedInTheName(identifierTree, symbolType) ||
+        typeWasMentionedInTheInitializer(initializer, symbolType))) {
+      reportIssue(identifierTree, MESSAGE);
+    }
+  }
+
+  private static boolean typeWasMentionedInTheName(IdentifierTree variable, Type type) {
+    return isLogicallyReferable(variable.name(), type.name());
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java:94]: You call this method many times with the same second argument, and end up doing multiple times the same operation: `type.name().toLowerCase(Locale.ROOT)`. We should compute it only once.

**Review by quentin-jaquier-sonarsource — COMMENTED** [review]
The current issues reported by the rule look good, I'm happy with the result!

Just one comment concerning the implementation.

**Review by quentin-jaquier-sonarsource — APPROVED** [review]
LGTM!

I'm quite happy with this implementation, looking forward to seeing it in action.

**Pull request merged**